File: SteamAPIWrap/Apps.cpp
#include "Precompiled.hpp"
#include "Apps.hpp"

#include "MemoryHelpers.h"

using namespace SteamAPIWrap;

namespace SteamAPIWrap
{
	template <> CApps * CSingleton<CApps>::instance = nullptr;

	CApps::CApps()
		: nativeDlcInstalledCallback(this, &CApps::OnDlcInstalled)
		, nativeRegisterActivationCodeResponseCallback(this, &CApps::OnRegisterActivationCodeResponse)
		, nativeAppProofOfPurchaseKeyResponseCallback(this, &CApps::OnAppProofOfPurchaseKeyResponse)
		, nativeNewLaunchQueryParametersCallback(this, &CApps::OnNewLaunchQueryParameters)
		, apps(nullptr)
	{
	}

	void CApps::SetSteamInterface(ISteamApps *apps)
	{
		this->apps = apps;
	}

	void CApps::RunCallbackSizeCheck()
	{
        DlcInstalled_t* dlcInstalled = new DlcInstalled_t();
		OnDlcInstalled(dlcInstalled);
		delete dlcInstalled;

        RegisterActivationCodeResponse_t* registerActivationCodeResponse = new RegisterActivationCodeResponse_t();
		OnRegisterActivationCodeResponse(registerActivationCodeResponse);
        delete registerActivationCodeResponse;

        AppProofOfPurchaseKeyResponse_t* appProofOfPurchaseKeyResponse = new AppProofOfPurchaseKeyResponse_t();
		OnAppProofOfPurchaseKeyResponse(appProofOfPurchaseKeyResponse);
        delete appProofOfPurchaseKeyResponse;

        NewLaunchQueryParameters_t* newLaunchQueryParameters = new NewLaunchQueryParameters_t();
		OnNewLaunchQueryParameters(newLaunchQueryParameters);
        delete newLaunchQueryParameters;
	}

	bool CApps::IsSubscribed()
	{
		return apps->BIsSubscribed();
	}

	bool CApps::IsLowViolence()
	{
		return apps->BIsLowViolence();
	}

	bool CApps::IsCybercafe()
	{
		return apps->BIsCybercafe();
	}

	bool CApps::IsVACBanned()
	{
		return apps->BIsVACBanned();
	}
	
	PConstantString CApps::GetCurrentGameLanguage()
	{
		return apps->GetCurrentGameLanguage();
	}

	PConstantString CApps::GetAvailableGameLanguages()
	{
		return apps->GetAvailableGameLanguages();
	}

	
	bool CApps::IsSubscribedApp( u32 appID )
	{
		return apps->BIsSubscribedApp( appID );
	}

	bool CApps::IsDlcInstalled( u32 appID )
	{
		return apps->BIsDlcInstalled( appID );
	}

	u32 CApps::GetEarliestPurchaseUnixTime( u32 appID )
	{
		return apps->GetEarliestPurchaseUnixTime( appID );
	}

	bool CApps::IsSubscribedFromFreeWeekend()
	{
		return apps->BIsSubscribedFromFreeWeekend();
	}

	int CApps::GetDLCCount()
	{
		return apps->GetDLCCount();
	}
	
	bool CApps::GetDLCDataByIndex( int iDLC, u32 *pAppID, bool *pbAvailable, PString pchName, int cchNameBufferSize )
	{
		return apps->BGetDLCDataByIndex(iDLC, pAppID, pbAvailable, pchName, cchNameBufferSize);
	}

	void CApps::InstallDLC( u32 appID )
	{
		apps->InstallDLC( appID );
	}

	void CApps::UninstallDLC( u32 appID )
	{
		apps->UninstallDLC( appID );
	}

	void CApps::RequestAppProofOfPurchaseKey( u32 appID )
	{
		apps->RequestAppProofOfPurchaseKey( appID );
	}

	bool CApps::GetCurrentBetaName( PString pchName, int cchNameBufferSize )
	{
		return apps->GetCurrentBetaName( pchName, cchNameBufferSize );
	}

	bool CApps::MarkContentCorrupt( bool bMissingFilesOnly )
	{
		return apps->MarkContentCorrupt( bMissingFilesOnly );
	}

	u32 CApps::GetInstalledDepots( u32 appID, u32 *pvecDepots, u32 cMaxDepots )
	{
		return apps->GetInstalledDepots( appID, pvecDepots, cMaxDepots );
	}

	SteamID CApps::GetAppOwner()
	{
		return apps->GetAppOwner().ConvertToUint64();
	}

	PConstantString CApps::GetLaunchQueryParam( PConstantString pchKey )
	{
		return apps->GetLaunchQueryParam( pchKey );
	}
};

ManagedSteam_API bool Apps_IsSubscribed()
{
	return CApps::Instance().IsSubscribed();
}

ManagedSteam_API bool Apps_IsLowViolence()
{
	return CApps::Instance().IsLowViolence();
}

ManagedSteam_API bool Apps_IsCybercafe()
{
	return CApps::Instance().IsCybercafe();
}

ManagedSteam_API bool Apps_IsVACBanned()
{
	return CApps::Instance().IsVACBanned();
}

ManagedSteam_API PConstantString Apps_GetCurrentGameLanguage()
{
	return CApps::Instance().GetCurrentGameLanguage();
}

ManagedSteam_API PConstantString Apps_GetAvailableGameLanguages()
{
	return CApps::Instance().GetAvailableGameLanguages();
}

ManagedSteam_API bool Apps_IsSubscribedApp( u32 appID )
{
	return CApps::Instance().IsSubscribedApp( appID );
}

ManagedSteam_API bool Apps_IsDlcInstalled( u32 appID )
{
	return CApps::Instance().IsDlcInstalled( appID );
}

ManagedSteam_API u32 Apps_GetEarliestPurchaseUnixTime( u32 appID )
{
	return CApps::Instance().GetEarliestPurchaseUnixTime( appID );
}

ManagedSteam_API bool Apps_IsSubscribedFromFreeWeekend()
{
	return CApps::Instance().IsSubscribedFromFreeWeekend();
}

ManagedSteam_API int Apps_GetDLCCount()
{
	return CApps::Instance().GetDLCCount();
}

ManagedSteam_API bool Apps_GetDLCDataByIndex( int iDLC, u32 *pAppID, bool *pbAvailable, PString pchName, int cchNameBufferSize )
{
	return CApps::Instance().GetDLCDataByIndex( iDLC, pAppID, pbAvailable, pchName, cchNameBufferSize );
}

ManagedSteam_API void Apps_InstallDLC( u32 appID )
{
	CApps::Instance().InstallDLC( appID );
}

ManagedSteam_API void Apps_UninstallDLC( u32 appID )
{
	CApps::Instance().UninstallDLC( appID );
}

ManagedSteam_API void Apps_RequestAppProofOfPurchaseKey( u32 appID )
{
	CApps::Instance().RequestAppProofOfPurchaseKey( appID );
}

ManagedSteam_API bool Apps_GetCurrentBetaName( PString pchName, int cchNameBufferSize )
{
	return CApps::Instance().GetCurrentBetaName( pchName, cchNameBufferSize );
}

ManagedSteam_API bool Apps_MarkContentCorrupt( bool bMissingFilesOnly )
{
	return CApps::Instance().MarkContentCorrupt( bMissingFilesOnly );
}

ManagedSteam_API u32 Apps_GetInstalledDepots( u32 appID, u32 *pvecDepots, u32 cMaxDepots )
{
	return CApps::Instance().GetInstalledDepots( appID, pvecDepots, cMaxDepots );
}

ManagedSteam_API SteamID Apps_GetAppOwner()
{
	return CApps::Instance().GetAppOwner();
}

PConstantString Apps_GetLaunchQueryParam( PConstantString pchKey )
{
	return CApps::Instance().GetLaunchQueryParam( pchKey );
}

File: ManagedSteam/Implementations/Apps.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Runtime.InteropServices;

using ManagedSteam.CallbackStructures;
using ManagedSteam.SteamTypes;
using ManagedSteam.Utility;

namespace ManagedSteam.Implementations
{
    class Apps : SteamService, IApps
    {
        private List<DlcInstalled> dlcInstalled;
        private List<RegisterActivationCodeResponse> registerActivationCodeResponse;
        private List<AppProofOfPurchaseKeyResponse> appProofOfPurchaseKeyResponse;
        private List<NewLaunchQueryParameters> newLaunchQueryParameters;

        public Apps()
        {
            dlcInstalled = new List<DlcInstalled>();
            registerActivationCodeResponse = new List<RegisterActivationCodeResponse>();
            appProofOfPurchaseKeyResponse = new List<AppProofOfPurchaseKeyResponse>();
            newLaunchQueryParameters = new List<NewLaunchQueryParameters>();

            Callbacks[CallbackID.DlcInstalled] = (data, dataSize) => dlcInstalled.Add(CallbackStructures.DlcInstalled.Create(data, dataSize));
            Callbacks[CallbackID.RegisterActivationCodeResponse] = (data, dataSize) => registerActivationCodeResponse.Add(CallbackStructures.RegisterActivationCodeResponse.Create(data, dataSize));
            Callbacks[CallbackID.AppProofOfPurchaseKeyResponse] = (data, dataSize) => appProofOfPurchaseKeyResponse.Add(CallbackStructures.AppProofOfPurchaseKeyResponse.Create(data, dataSize));
            Callbacks[CallbackID.NewLaunchQueryParameters] = (data, dataSize) => newLaunchQueryParameters.Add(CallbackStructures.NewLaunchQueryParameters.Create(data, dataSize));
        }

        public event CallbackEvent<DlcInstalled> DlcInstalled;
        public event CallbackEvent<RegisterActivationCodeResponse> RegisterActivationCodeResponse;
        public event CallbackEvent<AppProofOfPurchaseKeyResponse> AppProofOfPurchaseKeyResponse;
        public event CallbackEvent<NewLaunchQueryParameters> NewLaunchQueryParameters;

        internal override void CheckIfUsableInternal()
        {
            
        }

        internal override void ReleaseManagedResources()
        {
            dlcInstalled = null;
            DlcInstalled = null;
            registerActivationCodeResponse = null;
            RegisterActivationCodeResponse = null;
            appProofOfPurchaseKeyResponse = null;
            AppProofOfPurchaseKeyResponse = null;
            newLaunchQueryParameters = null;
            NewLaunchQueryParameters = null;
        }

        internal override void InvokeEvents()
        {
            InvokeEvents(dlcInstalled, DlcInstalled);
            InvokeEvents(registerActivationCodeResponse, RegisterActivationCodeResponse);
            InvokeEvents(appProofOfPurchaseKeyResponse, AppProofOfPurchaseKeyResponse);
            InvokeEvents(newLaunchQueryParameters, NewLaunchQueryParameters);
        }


        public bool IsSubscribed()
        {
            CheckIfUsable();

            return NativeMethods.Apps_IsSubscribed();
        }

        public bool IsLowViolence()
        {
            CheckIfUsable();

            return NativeMethods.Apps_IsLowViolence();
        }

        public bool IsCybercafe()
        {
            CheckIfUsable();

            return NativeMethods.Apps_IsCybercafe();
        }

        public bool IsVACBanned()
        {
            CheckIfUsable();

            return NativeMethods.Apps_IsVACBanned();
        }

        public string GetCurrentGameLanguage()
        {
            CheckIfUsable();

            return NativeHelpers.ToStringAnsi(NativeMethods.Apps_GetCurrentGameLanguage());
        }

        public string GetAvailableGameLanguages()
        {
            CheckIfUsable();

            return NativeHelpers.ToStringAnsi(NativeMethods.Apps_GetAvailableGameLanguages());
        }

        public bool IsSubscribedApp(AppID appID)
        {
            CheckIfUsable();
            return NativeMethods.Apps_IsSubscribedApp( appID.AsUInt32 );
        }

        public bool IsDlcInstalled(AppID appID)
        {
            CheckIfUsable();
            return NativeMethods.Apps_IsDlcInstalled( appID.AsUInt32 );
        }

        public uint GetEarliestPurchaseUnixTime(AppID appID)
        {
            CheckIfUsable();
            return NativeMethods.Apps_GetEarliestPurchaseUnixTime( appID.AsUInt32 );
        }

        public bool IsSubscribedFromFreeWeekend()
        {
            CheckIfUsable();
            return NativeMethods.Apps_IsSubscribedFromFreeWeekend();
        }

        public int GetDLCCount()
        {
            CheckIfUsable();
            return NativeMethods.Apps_GetDLCCount();
        }

        public bool GetDLCDataByIndex(int iDLC, out AppID pAppID, out bool pbAvailable, out string pchName)
        {
            CheckIfUsable();
            using (NativeBuffer buffer = new NativeBuffer(Constants.Apps.MaxAchievementNameLength))
            {
                uint localAppId = 0;
                pbAvailable = false;
                
                bool result = NativeMethods.Apps_GetDLCDataByIndex(iDLC, ref localAppId, ref pbAvailable, buffer.UnmanagedMemory, buffer.UnmanagedSize);
                pAppID = new AppID(localAppId);

                pchName = NativeHelpers.ToStringAnsi(buffer.UnmanagedMemory);

                return result;
            }
        }

        public AppsGetDLCDataByIndexResult GetDLCDataByIndex(int iDLC)
        {
            AppsGetDLCDataByIndexResult result = new AppsGetDLCDataByIndexResult();

            result.Result = GetDLCDataByIndex(iDLC, out result.AppID, out result.Available, out result.Name);

            return result;
        }

        public void InstallDLC(AppID appID)
        {
            CheckIfUsable();
            NativeMethods.Apps_InstallDLC( appID.AsUInt32 );
        }

        public void UninstallDLC(AppID appID)
        {
            CheckIfUsable();
            NativeMethods.Apps_UninstallDLC( appID.AsUInt32 );
        }

        public void RequestAppProofOfPurchaseKey(AppID appID)
        {
            CheckIfUsable();
            NativeMethods.Apps_RequestAppProofOfPurchaseKey( appID.AsUInt32 );
        }

        public bool GetCurrentBetaName(out string pchName)
        {
            CheckIfUsable();
            using (NativeBuffer buffer = new NativeBuffer(Constants.Apps.MaxBetaNameLength))
            {
                bool result = NativeMethods.Apps_GetCurrentBetaName(buffer.UnmanagedMemory, buffer.UnmanagedSize);
                pchName = NativeHelpers.ToStringAnsi(buffer.UnmanagedMemory);

                return result;
            }
        }

        public AppsGetCurrentBetaNameResult GetCurrentBetaName()
        {
            AppsGetCurrentBetaNameResult result = new AppsGetCurrentBetaNameResult();

            result.Result = GetCurrentBetaName(out result.Name);

            return result;
        }

        public bool MarkContentCorrupt(bool bMissingFilesOnly)
        {
            CheckIfUsable();
            return NativeMethods.Apps_MarkContentCorrupt( bMissingFilesOnly );
        }

        public uint GetInstalledDepots(AppID appID, out DepotID pDepotID, uint maxDepots)
        {
            CheckIfUsable();

            uint localDepotId = 0;

            uint result = NativeMethods.Apps_GetInstalledDepots(appID.AsUInt32, ref localDepotId, maxDepots);

            pDepotID = new DepotID(localDepotId);

            return result;
        }

        public SteamID GetAppOwner()
        {
            CheckIfUsable();

            return new SteamID(NativeMethods.Apps_GetAppOwner());
        }

        public string GetLaunchQueryParam(string key)
        {
            CheckIfUsable();

            return NativeHelpers.ToStringAnsi(NativeMethods.Apps_GetLaunchQueryParam(key));
        }
    }
}
